In [3]:
import pandas as pd
import numpy as np
import warnings
warnings.filterwarnings('ignore')

In [4]:
# Importing the dataset
df = pd.read_csv('lc_loan.csv')

In [5]:
# Getting informations about the dataset
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 887379 entries, 0 to 887378
Data columns (total 74 columns):
 #   Column                       Non-Null Count   Dtype  
---  ------                       --------------   -----  
 0   id                           887379 non-null  int64  
 1   member_id                    887379 non-null  int64  
 2   loan_amnt                    887379 non-null  float64
 3   funded_amnt                  887379 non-null  float64
 4   funded_amnt_inv              887379 non-null  float64
 5   term                         887379 non-null  object 
 6   int_rate                     887379 non-null  float64
 7   installment                  887379 non-null  float64
 8   grade                        887379 non-null  object 
 9   sub_grade                    887379 non-null  object 
 10  emp_title                    835917 non-null  object 
 11  emp_length                   842554 non-null  object 
 12  home_ownership               887379 non-null  object 
 13 

In [6]:
# Selecting some few important features
df1 = df[['loan_amnt', 'term', 'int_rate', 'installment', 'grade', 'sub_grade',
          'emp_length', 'home_ownership', 'annual_inc', 'verification_status',
          'purpose', 'dti', 'delinq_2yrs', 'earliest_cr_line', 'inq_last_6mths',
          'mths_since_last_delinq', 'mths_since_last_record', 'open_acc',
          'pub_rec', 'revol_bal', 'revol_util', 'total_acc', 'initial_list_status',
          'out_prncp', 'total_pymnt', 'total_rec_prncp', 'total_rec_int',
          'total_rec_late_fee', 'recoveries', 'collection_recovery_fee',
          'last_pymnt_d', 'last_pymnt_amnt', 'collections_12_mths_ex_med',
          'mths_since_last_major_derog', 'application_type', 'acc_now_delinq',
          'tot_coll_amt', 'tot_cur_bal', 'total_rev_hi_lim', 'zip_code', 'addr_state','loan_status']]
df1.head()

loan_amnt        term  int_rate  installment grade sub_grade emp_length  \
0     5000.0   36 months     10.65       162.87     B        B2  10+ years   
1     2500.0   60 months     15.27        59.83     C        C4   < 1 year   
2     2400.0   36 months     15.96        84.33     C        C5  10+ years   
3    10000.0   36 months     13.49       339.31     C        C1  10+ years   
4     3000.0   60 months     12.69        67.79     B        B5     1 year   

  home_ownership  annual_inc verification_status  ...  \
0           RENT     24000.0            Verified  ...   
1           RENT     30000.0     Source Verified  ...   
2           RENT     12252.0        Not Verified  ...   
3           RENT     49200.0     Source Verified  ...   
4           RENT     80000.0     Source Verified  ...   

  collections_12_mths_ex_med  mths_since_last_major_derog  application_type  \
0                        0.0                          NaN        INDIVIDUAL   
1                        0.0                          NaN        INDIVIDUAL   
2                        0.0                          NaN        INDIVIDUAL   
3                        0.0                          NaN        INDIVIDUAL   
4                        0.0                          NaN        INDIVIDUAL   

  acc_now_delinq  tot_coll_amt  tot_cur_bal  total_rev_hi_lim  zip_code  \
0            0.0           NaN          NaN               NaN     860xx   
1            0.0           NaN          NaN               NaN     309xx   
2            0.0           NaN          NaN               NaN     606xx   
3            0.0           NaN          NaN               NaN     917xx   
4            0.0           NaN          NaN               NaN     972xx   

   addr_state  loan_status  
0          AZ   Fully Paid  
1          GA  Charged Off  
2          IL   Fully Paid  
3          CA   Fully Paid  
4          OR      Current  

[5 rows x 42 columns]

In [7]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 887379 entries, 0 to 887378
Data columns (total 42 columns):
 #   Column                       Non-Null Count   Dtype  
---  ------                       --------------   -----  
 0   loan_amnt                    887379 non-null  float64
 1   term                         887379 non-null  object 
 2   int_rate                     887379 non-null  float64
 3   installment                  887379 non-null  float64
 4   grade                        887379 non-null  object 
 5   sub_grade                    887379 non-null  object 
 6   emp_length                   842554 non-null  object 
 7   home_ownership               887379 non-null  object 
 8   annual_inc                   887375 non-null  float64
 9   verification_status          887379 non-null  object 
 10  purpose                      887379 non-null  object 
 11  dti                          887379 non-null  float64
 12  delinq_2yrs                  887350 non-null  float64
 13 

In [8]:
# Checking for missing value in the dataset
df1.isnull().sum()

loan_amnt                           0
term                                0
int_rate                            0
installment                         0
grade                               0
sub_grade                           0
emp_length                      44825
home_ownership                      0
annual_inc                          4
verification_status                 0
purpose                             0
dti                                 0
delinq_2yrs                        29
earliest_cr_line                   29
inq_last_6mths                     29
mths_since_last_delinq         454312
mths_since_last_record         750326
open_acc                           29
pub_rec                            29
revol_bal                           0
revol_util                        502
total_acc                          29
initial_list_status                 0
out_prncp                           0
total_pymnt                         0
total_rec_prncp                     0
total_rec_in

In [9]:
# Dropping some columns with huge number of missing values
df1.drop(['mths_since_last_delinq', 'mths_since_last_record', 'mths_since_last_major_derog',  'tot_coll_amt', 'tot_cur_bal', 'total_rev_hi_lim','last_pymnt_d',  'emp_length'], axis=1, inplace=True)

In [10]:
# Checking for missing value in the dataset
df1.isnull().sum()

loan_amnt                       0
term                            0
int_rate                        0
installment                     0
grade                           0
sub_grade                       0
home_ownership                  0
annual_inc                      4
verification_status             0
purpose                         0
dti                             0
delinq_2yrs                    29
earliest_cr_line               29
inq_last_6mths                 29
open_acc                       29
pub_rec                        29
revol_bal                       0
revol_util                    502
total_acc                      29
initial_list_status             0
out_prncp                       0
total_pymnt                     0
total_rec_prncp                 0
total_rec_int                   0
total_rec_late_fee              0
recoveries                      0
collection_recovery_fee         0
last_pymnt_amnt                 0
collections_12_mths_ex_med    145
application_ty

In [11]:
# Filling up missing value with the mean value
df1['revol_util'] = df1['revol_util'].fillna(df1['revol_util'].mean())

In [12]:
# Filling up missing value with the mean value
df1['collections_12_mths_ex_med'] = df1['collections_12_mths_ex_med'].fillna(df1['collections_12_mths_ex_med'].mean())

In [13]:
# Summary statistics for the columns
df1.describe().T

count          mean           std     min  \
loan_amnt                   887379.0  14755.264605   8435.455601  500.00   
int_rate                    887379.0     13.246740      4.381867    5.32   
installment                 887379.0    436.717127    244.186593   15.67   
annual_inc                  887375.0  75027.587761  64698.300142    0.00   
dti                         887379.0     18.157039     17.190626    0.00   
delinq_2yrs                 887350.0      0.314442      0.862244    0.00   
inq_last_6mths              887350.0      0.694623      0.998448    0.00   
open_acc                    887350.0     11.548469      5.317313    0.00   
pub_rec                     887350.0      0.195307      0.582091    0.00   
revol_bal                   887379.0  16920.787533  22426.791896    0.00   
revol_util                  887379.0     55.067693     23.827601    0.00   
total_acc                   887350.0     25.268026     11.840561    1.00   
out_prncp                   887379.0   8403.124852   8493.450269    0.00   
total_pymnt                 887379.0   7558.826684   7871.243336    0.00   
total_rec_prncp             887379.0   5757.706424   6625.441046    0.00   
total_rec_int               887379.0   1754.804327   2095.356099    0.00   
total_rec_late_fee          887379.0      0.396692      4.087825    0.00   
recoveries                  887379.0     45.919243    409.693874    0.00   
collection_recovery_fee     887379.0      4.880757     63.125281    0.00   
last_pymnt_amnt             887379.0   2164.145585   4794.783233    0.00   
collections_12_mths_ex_med  887379.0      0.014380      0.134180    0.00   
acc_now_delinq              887350.0      0.004991      0.077625    0.00   

                                  25%           50%           75%  \
loan_amnt                    8000.000  13000.000000  20000.000000   
int_rate                        9.990     12.990000     16.200000   
installment                   260.705    382.550000    572.600000   
annual_inc                  45000.000  65000.000000  90000.000000   
dti                            11.910     17.650000     23.950000   
delinq_2yrs                     0.000      0.000000      0.000000   
inq_last_6mths                  0.000      0.000000      1.000000   
open_acc                        8.000     11.000000     14.000000   
pub_rec                         0.000      0.000000      0.000000   
revol_bal                    6443.000  11875.000000  20829.000000   
revol_util                     37.700     56.000000     73.600000   
total_acc                      17.000     24.000000     32.000000   
out_prncp                       0.000   6458.290000  13659.055000   
total_pymnt                  1914.590   4894.999117  10616.814231   
total_rec_prncp              1200.570   3215.320000   8000.000000   
total_rec_int                 441.460   1073.280000   2238.265000   
total_rec_late_fee              0.000      0.000000      0.000000   
recoveries                      0.000      0.000000      0.000000   
collection_recovery_fee         0.000      0.000000      0.000000   
last_pymnt_amnt               280.170    462.780000    831.245000   
collections_12_mths_ex_med      0.000      0.000000      0.000000   
acc_now_delinq                  0.000      0.000000      0.000000   

                                     max  
loan_amnt                   3.500000e+04  
int_rate                    2.899000e+01  
installment                 1.445460e+03  
annual_inc                  9.500000e+06  
dti                         9.999000e+03  
delinq_2yrs                 3.900000e+01  
inq_last_6mths              3.300000e+01  
open_acc                    9.000000e+01  
pub_rec                     8.600000e+01  
revol_bal                   2.904836e+06  
revol_util                  8.923000e+02  
total_acc                   1.690000e+02  
out_prncp                   4.937286e+04  
total_pymnt                 5.777758e+04  
total_rec_prncp             3.500003e+04  
total_

In [14]:
# Dropping missing rows and  assigning into a new dataframe
df2 = df1.dropna()

In [15]:
df2.info()

<class 'pandas.core.frame.DataFrame'>
Index: 887350 entries, 0 to 887378
Data columns (total 34 columns):
 #   Column                      Non-Null Count   Dtype  
---  ------                      --------------   -----  
 0   loan_amnt                   887350 non-null  float64
 1   term                        887350 non-null  object 
 2   int_rate                    887350 non-null  float64
 3   installment                 887350 non-null  float64
 4   grade                       887350 non-null  object 
 5   sub_grade                   887350 non-null  object 
 6   home_ownership              887350 non-null  object 
 7   annual_inc                  887350 non-null  float64
 8   verification_status         887350 non-null  object 
 9   purpose                     887350 non-null  object 
 10  dti                         887350 non-null  float64
 11  delinq_2yrs                 887350 non-null  float64
 12  earliest_cr_line            887350 non-null  object 
 13  inq_last_6mths     

In [16]:
# Getting the unique value of the loan status column
df2['loan_status'].unique()

array(['Fully Paid', 'Charged Off', 'Current', 'Default',
       'Late (31-120 days)', 'In Grace Period', 'Late (16-30 days)',
       'Does not meet the credit policy. Status:Fully Paid',
       'Does not meet the credit policy. Status:Charged Off', 'Issued'],
      dtype=object)

In [17]:
from sklearn.preprocessing import LabelEncoder, StandardScaler

In [18]:
# Encoding the categorical variables
label_encoders = {}
for column in df2.select_dtypes(include=['object']).columns:
    label_encoders[column] = LabelEncoder()
    df2[column] = label_encoders[column].fit_transform(df2[column])



In [20]:
# Verifing that there are no more missing values.
df2.isnull().sum()

loan_amnt                     0
term                          0
int_rate                      0
installment                   0
grade                         0
sub_grade                     0
home_ownership                0
annual_inc                    0
verification_status           0
purpose                       0
dti                           0
delinq_2yrs                   0
earliest_cr_line              0
inq_last_6mths                0
open_acc                      0
pub_rec                       0
revol_bal                     0
revol_util                    0
total_acc                     0
initial_list_status           0
out_prncp                     0
total_pymnt                   0
total_rec_prncp               0
total_rec_int                 0
total_rec_late_fee            0
recoveries                    0
collection_recovery_fee       0
last_pymnt_amnt               0
collections_12_mths_ex_med    0
application_type              0
acc_now_delinq                0
zip_code

In [21]:
# Checking for relationship between the columns
correlation_matrix = df2.corr()

In [22]:
correlation_with_target = correlation_matrix['loan_status'].drop('loan_status').sort_values(ascending=False)
correlation_with_target

last_pymnt_amnt               0.511136
total_rec_prncp               0.491828
total_pymnt                   0.416439
inq_last_6mths                0.076561
purpose                       0.050593
total_rec_late_fee            0.040874
total_rec_int                 0.031706
int_rate                      0.029161
zip_code                      0.018318
home_ownership                0.003481
total_acc                     0.001040
earliest_cr_line              0.000272
application_type             -0.003457
annual_inc                   -0.005150
acc_now_delinq               -0.009229
verification_status          -0.011078
sub_grade                    -0.014760
grade                        -0.014906
addr_state                   -0.016173
collections_12_mths_ex_med   -0.022160
delinq_2yrs                  -0.025955
revol_util                   -0.033714
pub_rec                      -0.034432
revol_bal                    -0.035012
installment                  -0.035330
dti                      

In [23]:
# Assigning X and y variables
X = df2.drop('loan_status', axis=1)
y = df2['loan_status']

In [24]:
from sklearn.feature_selection import mutual_info_classif

In [25]:
# Checking for dependency between the target and features
mi_scores = mutual_info_classif(X, y, discrete_features='auto')
mi_scores = pd.Series(mi_scores, name='MI Scores', index=X.columns)
mi_scores = mi_scores.sort_values(ascending=False)

In [26]:
mi_scores

out_prncp                     0.634054
total_rec_prncp               0.588891
last_pymnt_amnt               0.466380
total_pymnt                   0.277239
int_rate                      0.200756
installment                   0.138703
total_rec_int                 0.138073
initial_list_status           0.125793
recoveries                    0.091998
collection_recovery_fee       0.086755
verification_status           0.049574
home_ownership                0.048018
term                          0.039978
purpose                       0.033947
grade                         0.031829
inq_last_6mths                0.029226
sub_grade                     0.018822
loan_amnt                     0.018738
dti                           0.013036
total_rec_late_fee            0.011305
addr_state                    0.008847
open_acc                      0.008800
annual_inc                    0.006920
earliest_cr_line              0.005686
revol_util                    0.005286
revol_bal                

In [27]:
# Selected columns as features after evaluation of the correlation and mutual info classification
selected_columns = [
    'last_pymnt_amnt', 'total_rec_prncp', 'total_pymnt', 'int_rate', 'total_rec_int',
    'out_prncp', 'inq_last_6mths', 'purpose', 'initial_list_status', 'recoveries',
    'collection_recovery_fee'
]

In [28]:
X= df2[selected_columns]

In [29]:
from sklearn.model_selection import train_test_split

In [32]:
# Splitting the dataset for modelling
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

In [30]:
from sklearn.tree import DecisionTreeClassifier

In [33]:
# Decision tree classifier - initial
clf = DecisionTreeClassifier(random_state=42)
clf.fit(X_train, y_train)

DecisionTreeClassifier(random_state=42)

In [34]:
# Predicting the target variable on the testing data.
y_pred = clf.predict(X_test)

In [35]:
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix

In [36]:
# Evaluating the model performance
importances = clf.feature_importances_
feature_names = X.columns
features_df = pd.DataFrame({'Feature': feature_names, 'Importance': importances})

In [37]:
features_df.sort_values(by='Importance', ascending=False)

Feature  Importance
5                 out_prncp    0.710686
9                recoveries    0.084532
1           total_rec_prncp    0.063524
0           last_pymnt_amnt    0.044494
4             total_rec_int    0.039218
3                  int_rate    0.019982
2               total_pymnt    0.015267
6            inq_last_6mths    0.010305
7                   purpose    0.008411
8       initial_list_status    0.003332
10  collection_recovery_fee    0.000249

In [38]:
label = ['Fully Paid', 'Charged Off', 'Current', 'Default',
       'Late (31-120 days)', 'In Grace Period', 'Late (16-30 days)',
       'Does not meet the credit policy. Status:Fully Paid',
       'Does not meet the credit policy. Status:Charged Off', 'Issued']
pd.DataFrame(confusion_matrix(y_test, y_pred), columns= label, index=label)

Fully Paid  Charged Off  \
Fully Paid                                                8894            0   
Charged Off                                                  0       116556   
Current                                                      0          206   
Default                                                     65            0   
Late (31-120 days)                                           4            0   
In Grace Period                                            161           17   
Late (16-30 days)                                            0         1152   
Does not meet the credit policy. Status:Fully Paid           0          627   
Does not meet the credit policy. Status:Charged...           0          396   
Issued                                                       0         1824   

                                                    Current  Default  \
Fully Paid                                                0       94   
Charged Off                                             198        0   
Current                                                   7        0   
Default                                                   0       76   
Late (31-120 days)                                        0        4   
In Grace Period                                           0        4   
Late (16-30 days)                                         6        0   
Does not meet the credit policy. Status:Fully Paid        1        0   
Does not meet the credit policy. Status:Charged...        3        0   
Issued                                                   30        0   

                                                    Late (31-120 days)  \
Fully Paid                                                           9   
Charged Off                                                          0   
Current                                                              0   
Default                                                              5   
Late (31-120 days)                                                 178   
In Grace Period                                                    200   
Late (16-30 days)                                                    0   
Does not meet the credit policy. Status:Fully Paid                   0   
Does not meet the credit policy. Status:Charged...                   0   
Issued                                                               0   

                                                    In Grace Period  \
Fully Paid                                                      177   
Charged Off                                                      13   
Current                                                           0   
Default                                                           1   
Late (31-120 days)                                              223   
In Grace Period                                               40941   
Late (16-30 days)                                                 1   
Does not meet the credit policy. Status:Fully Paid                1   
Does not meet the credit policy. Status:Charged...                0   
Issued                                                            0   

                                                    Late (16-30 days)  \
Fully Paid                                                          0   
Charged Off                                                      1068   
Current                                                             6   
Default                                                             0   
Late (31-120 days)                                                  0   
In Grace Period                                                     2   
Late (16-30 days)                                                  35   
Does not meet the credit policy. Status:Fully Paid                 14   
Does not meet the credit policy. Status:Charged...                 10   
Issued                                                        

In [39]:
# Evaluating the model performance
print(classification_report(y_test, y_pred, target_names=label))

                                                     precision    recall  f1-score   support

                                         Fully Paid       0.97      0.97      0.97      9174
                                        Charged Off       0.97      0.97      0.97    120471
                                            Current       0.03      0.03      0.03       267
                                            Default       0.43      0.52      0.47       147
                                 Late (31-120 days)       0.45      0.44      0.44       409
                                    In Grace Period       0.99      0.99      0.99     41327
                                  Late (16-30 days)       0.03      0.03      0.03      1267
 Does not meet the credit policy. Status:Fully Paid       0.70      0.62      0.66      1726
Does not meet the credit policy. Status:Charged Off       0.02      0.02      0.02       441
                                             Issued       0.12      0

In [40]:
# Decision classifier with parameter tuned
clf = DecisionTreeClassifier(max_depth=8, random_state=42)
clf.fit(X_train, y_train)

DecisionTreeClassifier(max_depth=8, random_state=42)

In [41]:
# Predicting the target variable on the testing data.
y_pred = clf.predict(X_test)

In [42]:
# Evaluating the model performance
importances = clf.feature_importances_
feature_names = X.columns
features_df = pd.DataFrame({'Feature': feature_names, 'Importance': importances})
features_df.sort_values(by='Importance', ascending=False)

Feature  Importance
5                 out_prncp    0.789935
9                recoveries    0.096907
1           total_rec_prncp    0.052894
0           last_pymnt_amnt    0.030198
4             total_rec_int    0.021710
3                  int_rate    0.004402
6            inq_last_6mths    0.003082
8       initial_list_status    0.000533
2               total_pymnt    0.000241
10  collection_recovery_fee    0.000070
7                   purpose    0.000027

In [43]:
# Evaluating the model performance
pd.DataFrame(confusion_matrix(y_test, y_pred), columns= label, index=label)

Fully Paid  Charged Off  \
Fully Paid                                                8577            0   
Charged Off                                                  0       119240   
Current                                                      0          267   
Default                                                     73            0   
Late (31-120 days)                                           3            0   
In Grace Period                                            232            0   
Late (16-30 days)                                            0         1237   
Does not meet the credit policy. Status:Fully Paid           0          206   
Does not meet the credit policy. Status:Charged...           0          437   
Issued                                                       0         2210   

                                                    Current  Default  \
Fully Paid                                                0       31   
Charged Off                                               0        0   
Current                                                   0        0   
Default                                                   0       67   
Late (31-120 days)                                        0        1   
In Grace Period                                           0        0   
Late (16-30 days)                                         0        0   
Does not meet the credit policy. Status:Fully Paid        0        0   
Does not meet the credit policy. Status:Charged...        0        0   
Issued                                                    0        0   

                                                    Late (31-120 days)  \
Fully Paid                                                           8   
Charged Off                                                          0   
Current                                                              0   
Default                                                              3   
Late (31-120 days)                                                 114   
In Grace Period                                                     54   
Late (16-30 days)                                                    0   
Does not meet the credit policy. Status:Fully Paid                   0   
Does not meet the credit policy. Status:Charged...                   0   
Issued                                                               0   

                                                    In Grace Period  \
Fully Paid                                                      558   
Charged Off                                                      13   
Current                                                           0   
Default                                                           4   
Late (31-120 days)                                              291   
In Grace Period                                               41041   
Late (16-30 days)                                                 1   
Does not meet the credit policy. Status:Fully Paid                1   
Does not meet the credit policy. Status:Charged...                0   
Issued                                                            0   

                                                    Late (16-30 days)  \
Fully Paid                                                          0   
Charged Off                                                         0   
Current                                                             0   
Default                                                             0   
Late (31-120 days)                                                  0   
In Grace Period                                                     0   
Late (16-30 days)                                                   0   
Does not meet the credit policy. Status:Fully Paid                  0   
Does not meet the credit policy. Status:Charged...                  0   
Issued                                                        

In [44]:
# Evaluating the model performance
print(classification_report(y_test, y_pred, target_names=label))

                                                     precision    recall  f1-score   support

                                         Fully Paid       0.97      0.93      0.95      9174
                                        Charged Off       0.96      0.99      0.98    120471
                                            Current       0.00      0.00      0.00       267
                                            Default       0.68      0.46      0.54       147
                                 Late (31-120 days)       0.64      0.28      0.39       409
                                    In Grace Period       0.98      0.99      0.99     41327
                                  Late (16-30 days)       0.00      0.00      0.00      1267
 Does not meet the credit policy. Status:Fully Paid       0.55      0.88      0.68      1726
Does not meet the credit policy. Status:Charged Off       0.33      0.00      0.00       441
                                             Issued       0.68      0

In [45]:
from sklearn.ensemble import RandomForestClassifier

In [46]:
# Random forest modelling - initial
forest = RandomForestClassifier(n_estimators= 100, random_state=42)
forest.fit(X_train, y_train)

RandomForestClassifier(random_state=42)

In [48]:
# Predicting the target variable on the testing data.
forest_pred = forest.predict(X_test)

In [49]:
# Evaluating the model performance
pd.DataFrame(confusion_matrix(y_test, forest_pred), columns= label, index=label)

Fully Paid  Charged Off  \
Fully Paid                                                8979            0   
Charged Off                                                  0       119657   
Current                                                      0          265   
Default                                                     71            0   
Late (31-120 days)                                           4            0   
In Grace Period                                             45            0   
Late (16-30 days)                                            0         1248   
Does not meet the credit policy. Status:Fully Paid           0          624   
Does not meet the credit policy. Status:Charged...           0          435   
Issued                                                       0         2186   

                                                    Current  Default  \
Fully Paid                                                0       13   
Charged Off                                               7        0   
Current                                                   1        0   
Default                                                   0       66   
Late (31-120 days)                                        0        1   
In Grace Period                                           0        0   
Late (16-30 days)                                         0        0   
Does not meet the credit policy. Status:Fully Paid        0        0   
Does not meet the credit policy. Status:Charged...        0        0   
Issued                                                    1        0   

                                                    Late (31-120 days)  \
Fully Paid                                                           1   
Charged Off                                                          0   
Current                                                              0   
Default                                                              8   
Late (31-120 days)                                                 161   
In Grace Period                                                     39   
Late (16-30 days)                                                    0   
Does not meet the credit policy. Status:Fully Paid                   0   
Does not meet the credit policy. Status:Charged...                   0   
Issued                                                               0   

                                                    In Grace Period  \
Fully Paid                                                      181   
Charged Off                                                      16   
Current                                                           0   
Default                                                           2   
Late (31-120 days)                                              243   
In Grace Period                                               41243   
Late (16-30 days)                                                 2   
Does not meet the credit policy. Status:Fully Paid                1   
Does not meet the credit policy. Status:Charged...                0   
Issued                                                            0   

                                                    Late (16-30 days)  \
Fully Paid                                                          0   
Charged Off                                                       121   
Current                                                             0   
Default                                                             0   
Late (31-120 days)                                                  0   
In Grace Period                                                     0   
Late (16-30 days)                                                   1   
Does not meet the credit policy. Status:Fully Paid                  8   
Does not meet the credit policy. Status:Charged...                  0   
Issued                                                        

In [50]:
# Evaluating the model performance

print(classification_report(y_test, forest_pred, target_names=label))

                                                     precision    recall  f1-score   support

                                         Fully Paid       0.99      0.98      0.98      9174
                                        Charged Off       0.96      0.99      0.98    120471
                                            Current       0.11      0.00      0.01       267
                                            Default       0.82      0.45      0.58       147
                                 Late (31-120 days)       0.77      0.39      0.52       409
                                    In Grace Period       0.99      1.00      0.99     41327
                                  Late (16-30 days)       0.01      0.00      0.00      1267
 Does not meet the credit policy. Status:Fully Paid       0.67      0.63      0.65      1726
Does not meet the credit policy. Status:Charged Off       0.00      0.00      0.00       441
                                             Issued       0.28      0

In [51]:
# Randon forest classifier with hyperparameter tuned
forest = RandomForestClassifier(n_estimators= 100, random_state=42, max_depth=7)
forest.fit(X_train, y_train)

RandomForestClassifier(max_depth=7, random_state=42)

In [52]:
# Predicting the target variable on the testing data.
forest_pred = forest.predict(X_test)

In [53]:
# Evaluating the model performance
pd.DataFrame(confusion_matrix(y_test, forest_pred), columns= label, index=label)

Fully Paid  Charged Off  \
Fully Paid                                                8197            0   
Charged Off                                                  0       119487   
Current                                                      0          267   
Default                                                    113            0   
Late (31-120 days)                                          10            0   
In Grace Period                                            162            0   
Late (16-30 days)                                            0         1240   
Does not meet the credit policy. Status:Fully Paid           0          460   
Does not meet the credit policy. Status:Charged...           0          438   
Issued                                                       0         2220   

                                                    Current  Default  \
Fully Paid                                                0        1   
Charged Off                                               0        0   
Current                                                   0        0   
Default                                                   0       18   
Late (31-120 days)                                        0        0   
In Grace Period                                           0        0   
Late (16-30 days)                                         0        0   
Does not meet the credit policy. Status:Fully Paid        0        0   
Does not meet the credit policy. Status:Charged...        0        0   
Issued                                                    0        0   

                                                    Late (31-120 days)  \
Fully Paid                                                           0   
Charged Off                                                          0   
Current                                                              0   
Default                                                              0   
Late (31-120 days)                                                   0   
In Grace Period                                                      0   
Late (16-30 days)                                                    0   
Does not meet the credit policy. Status:Fully Paid                   0   
Does not meet the credit policy. Status:Charged...                   0   
Issued                                                               0   

                                                    In Grace Period  \
Fully Paid                                                      976   
Charged Off                                                      13   
Current                                                           0   
Default                                                          16   
Late (31-120 days)                                              399   
In Grace Period                                               41165   
Late (16-30 days)                                                 1   
Does not meet the credit policy. Status:Fully Paid                1   
Does not meet the credit policy. Status:Charged...                0   
Issued                                                            0   

                                                    Late (16-30 days)  \
Fully Paid                                                          0   
Charged Off                                                         0   
Current                                                             0   
Default                                                             0   
Late (31-120 days)                                                  0   
In Grace Period                                                     0   
Late (16-30 days)                                                   0   
Does not meet the credit policy. Status:Fully Paid                  0   
Does not meet the credit policy. Status:Charged...                  0   
Issued                                                        

In [54]:
# Evaluating the model performance
print(classification_report(y_test, forest_pred, target_names=label))

                                                     precision    recall  f1-score   support

                                         Fully Paid       0.97      0.89      0.93      9174
                                        Charged Off       0.96      0.99      0.98    120471
                                            Current       0.00      0.00      0.00       267
                                            Default       0.95      0.12      0.22       147
                                 Late (31-120 days)       0.00      0.00      0.00       409
                                    In Grace Period       0.97      1.00      0.98     41327
                                  Late (16-30 days)       0.00      0.00      0.00      1267
 Does not meet the credit policy. Status:Fully Paid       0.56      0.73      0.63      1726
Does not meet the credit policy. Status:Charged Off       0.00      0.00      0.00       441
                                             Issued       1.00      0

In [55]:
from sklearn.ensemble import GradientBoostingClassifier

In [56]:
# Gradient boosting classifier - initial
gradient = GradientBoostingClassifier( random_state=42)
gradient.fit(X_train, y_train)

GradientBoostingClassifier(random_state=42)

In [57]:
# Predicting the target variable on the testing data.
gradient_pred = gradient.predict(X_test)

In [58]:
# Evaluating the model performance
pd.DataFrame(confusion_matrix(y_test, gradient_pred), columns= label, index=label)

Fully Paid  Charged Off  \
Fully Paid                                                8692            0   
Charged Off                                                  0       119889   
Current                                                      0          266   
Default                                                     87            0   
Late (31-120 days)                                           3            0   
In Grace Period                                             80            0   
Late (16-30 days)                                            0         1254   
Does not meet the credit policy. Status:Fully Paid           0          552   
Does not meet the credit policy. Status:Charged...           0          438   
Issued                                                       0         2199   

                                                    Current  Default  \
Fully Paid                                                0       27   
Charged Off                                              22        0   
Current                                                   1        0   
Default                                                   0       42   
Late (31-120 days)                                        0       15   
In Grace Period                                           0        1   
Late (16-30 days)                                         0        0   
Does not meet the credit policy. Status:Fully Paid        0        0   
Does not meet the credit policy. Status:Charged...        0        0   
Issued                                                    0        0   

                                                    Late (31-120 days)  \
Fully Paid                                                          22   
Charged Off                                                          0   
Current                                                              0   
Default                                                             15   
Late (31-120 days)                                                 140   
In Grace Period                                                     22   
Late (16-30 days)                                                    0   
Does not meet the credit policy. Status:Fully Paid                   0   
Does not meet the credit policy. Status:Charged...                   0   
Issued                                                               0   

                                                    In Grace Period  \
Fully Paid                                                      432   
Charged Off                                                      13   
Current                                                           0   
Default                                                           3   
Late (31-120 days)                                              251   
In Grace Period                                               41222   
Late (16-30 days)                                                 1   
Does not meet the credit policy. Status:Fully Paid                1   
Does not meet the credit policy. Status:Charged...                0   
Issued                                                            0   

                                                    Late (16-30 days)  \
Fully Paid                                                          0   
Charged Off                                                         6   
Current                                                             0   
Default                                                             0   
Late (31-120 days)                                                  0   
In Grace Period                                                     1   
Late (16-30 days)                                                   0   
Does not meet the credit policy. Status:Fully Paid                  0   
Does not meet the credit policy. Status:Charged...                  0   
Issued                                                        

In [59]:
# Evaluating the model performance
print(classification_report(y_test, gradient_pred, target_names=label))

                                                     precision    recall  f1-score   support

                                         Fully Paid       0.98      0.95      0.96      9174
                                        Charged Off       0.96      1.00      0.98    120471
                                            Current       0.04      0.00      0.01       267
                                            Default       0.49      0.29      0.36       147
                                 Late (31-120 days)       0.70      0.34      0.46       409
                                    In Grace Period       0.98      1.00      0.99     41327
                                  Late (16-30 days)       0.00      0.00      0.00      1267
 Does not meet the credit policy. Status:Fully Paid       0.70      0.68      0.69      1726
Does not meet the credit policy. Status:Charged Off       0.02      0.00      0.00       441
                                             Issued       0.77      0

In [60]:
# Scaling the featured coulmns
scaled = StandardScaler()
X_scaled = scaled.fit_transform(X)

In [61]:
# Splitting the dataset for modelling
X_train, X_test, y_train, y_test = train_test_split(Xscaled, y, test_size=0.2, random_state=42)

In [62]:
from sklearn.linear_model import LogisticRegression


In [64]:
# Logistic regression - intial
logistic = LogisticRegression(random_state=42)
logistic.fit(X_train, y_train)

LogisticRegression(random_state=42)

In [65]:
# Predicting the target variable on the testing data.
y_pred = logistic.predict(X_test)

In [66]:
# Evaluating the model performance
pd.DataFrame(confusion_matrix(y_test, y_pred), columns= label, index=label)

Fully Paid  Charged Off  \
Fully Paid                                                8480            0   
Charged Off                                                117       119688   
Current                                                      1          266   
Default                                                     82            0   
Late (31-120 days)                                          18            0   
In Grace Period                                            401            0   
Late (16-30 days)                                            2         1245   
Does not meet the credit policy. Status:Fully Paid           0         1336   
Does not meet the credit policy. Status:Charged...           0          435   
Issued                                                       0         2228   

                                                    Current  Default  \
Fully Paid                                                0       12   
Charged Off                                               0        4   
Current                                                   0        0   
Default                                                   0       39   
Late (31-120 days)                                        0        6   
In Grace Period                                           0        0   
Late (16-30 days)                                         0        0   
Does not meet the credit policy. Status:Fully Paid        0        0   
Does not meet the credit policy. Status:Charged...        0        0   
Issued                                                    0        0   

                                                    Late (31-120 days)  \
Fully Paid                                                           9   
Charged Off                                                          6   
Current                                                              0   
Default                                                             11   
Late (31-120 days)                                                  88   
In Grace Period                                                     23   
Late (16-30 days)                                                    0   
Does not meet the credit policy. Status:Fully Paid                   0   
Does not meet the credit policy. Status:Charged...                   0   
Issued                                                               0   

                                                    In Grace Period  \
Fully Paid                                                      673   
Charged Off                                                     315   
Current                                                           0   
Default                                                          15   
Late (31-120 days)                                              297   
In Grace Period                                               40903   
Late (16-30 days)                                                 9   
Does not meet the credit policy. Status:Fully Paid                1   
Does not meet the credit policy. Status:Charged...                3   
Issued                                                            6   

                                                    Late (16-30 days)  \
Fully Paid                                                          0   
Charged Off                                                         0   
Current                                                             0   
Default                                                             0   
Late (31-120 days)                                                  0   
In Grace Period                                                     0   
Late (16-30 days)                                                   0   
Does not meet the credit policy. Status:Fully Paid                  0   
Does not meet the credit policy. Status:Charged...                  0   
Issued                                                        

In [67]:
# Evaluating the model performance
print(classification_report(y_test, y_pred, target_names=label))

                                                     precision    recall  f1-score   support

                                         Fully Paid       0.93      0.92      0.93      9174
                                        Charged Off       0.96      0.99      0.97    120471
                                            Current       0.00      0.00      0.00       267
                                            Default       0.64      0.27      0.38       147
                                 Late (31-120 days)       0.64      0.22      0.32       409
                                    In Grace Period       0.97      0.99      0.98     41327
                                  Late (16-30 days)       0.00      0.00      0.00      1267
 Does not meet the credit policy. Status:Fully Paid       0.52      0.23      0.31      1726
Does not meet the credit policy. Status:Charged Off       0.00      0.00      0.00       441
                                             Issued       0.00      0

### Summary of Model Results and Recommendations

**Decision Tree Classifier:**
- **Without Tuning:** High accuracy (0.95) but poor performance for minority classes (Macro Avg F1-Score: 0.47).
- **With Tuning:** Slight improvement in accuracy (0.96) but minimal change in handling minority classes (Macro Avg F1-Score: 0.45).

**Random Forest Classifier:**
- **Without Tuning:** High accuracy (0.96) with better handling of minority classes compared to the decision tree (Macro Avg F1-Score: 0.48).
- **With Tuning:** Accuracy remains high (0.96) but performance on minority classes worsens (Macro Avg F1-Score: 0.37).

**Gradient Boosting Classifier:**
- **Accuracy:** 0.96
- **Macro Avg F1-Score:** 0.45
- Balanced performance with relatively good handling of minority classes.

**Logistic Regression:**
- **Accuracy:** 0.96
- **Macro Avg F1-Score:** 0.39
- High accuracy but struggles significantly with minority classes.

**Recommended Model: Gradient Boosting Classifier**
- Offers a balanced performance with high accuracy and reasonable handling of minority classes.
- Best suited for predicting loan defaults due to its ability to manage non-linear relationships and interactions between features effectively.

### Performance Comparison: P2P Lending Models vs. Traditional Bank Lending Practices

**Peer-to-Peer (P2P) Lending Models:**

- **Accessibility and Flexibility:** P2P lending platforms provide more accessible and flexible options for borrowers, particularly for those who might find it difficult to secure loans from traditional banks due to stringent credit requirements. These platforms utilize alternative data and innovative credit assessment techniques to evaluate borrowers (Finance Magnates, 2023).
- **Interest Rates and Returns:** P2P lending often offers competitive interest rates for borrowers and higher returns for lenders compared to traditional savings accounts or fixed deposits. However, the risk of default is generally higher due to less stringent regulatory oversight (Fintech Basics, 2023).
- **Efficiency and Speed:** The use of automation and digital processes in P2P lending platforms streamlines loan origination and approval, making the process faster and more efficient (LiquiLoans, 2023).
- **Disintermediation:** By eliminating traditional intermediaries like banks, P2P platforms reduce overhead costs and can offer more attractive rates to both borrowers and lenders (Monexo, 2023).

**Traditional Bank Lending Practices:**

- **Security and Regulation:** Traditional banks are heavily regulated, providing a higher level of security for both borrowers and lenders. This regulation helps mitigate risks but can also make the lending process more cumbersome and slower (Fintech Basics, 2023).
- **Comprehensive Financial Services:** Banks offer a wide range of financial services beyond lending, such as savings accounts, investment products, and financial advice. This comprehensive approach can be beneficial for customers seeking a one-stop financial solution (Monexo, 2023).
- **Lower Default Risk:** Due to stricter credit assessments and regulatory requirements, traditional banks generally experience lower default rates compared to P2P lending platforms (Fintech Basics, 2023).
- **Higher Interest Rates and Stricter Requirements:** Traditional bank loans often come with higher interest rates and more stringent borrowing requirements, which can limit access for some borrowers (Fintech Basics, 2023).

### Role of Deposit Insurance in Providing a Strategic Advantage to Commercial Banks

**Deposit Insurance:**

- **Safety and Confidence:** Deposit insurance provides a safety net for depositors, ensuring that their funds are protected up to a certain limit in case of bank failure. This protection fosters confidence among depositors, making them more likely to keep their money in insured banks (IADI, 2019).
- **Strategic Advantage:** Commercial banks benefit from deposit insurance as it attracts more deposits, providing them with a stable source of funds for lending. This advantage is not available to P2P lending platforms or stablecoin solutions, which do not typically offer deposit insurance (IADI, 2019).
- **Subsidy Effect:** Deposit insurance can be seen as a subsidy to incumbent commercial banks because it reduces the perceived risk for depositors, allowing banks to attract and retain deposits more easily. This implicit subsidy helps banks maintain a competitive edge over newer financial technologies like P2P lending and stablecoins (Oxford Academic, 2023).

**Comparison with P2P Lending and Stablecoin Solutions:**

- **P2P Lending:** P2P platforms do not offer deposit insurance, which can make them less attractive to risk-averse investors and depositors. The lack of insurance means that lenders on these platforms bear the full risk of borrower defaults (Statista, 2020).
- **Stablecoin Solutions:** Stablecoins, while designed to maintain a stable value, also lack deposit insurance. This absence can deter users who prioritize the safety of their funds. Regulatory frameworks for stablecoins are still evolving, and they do not yet offer the same level of protection as traditional bank deposits (Bank of England, 2023).

In summary, deposit insurance provides a significant strategic advantage to commercial banks by enhancing depositor confidence and attracting more stable funding sources. This advantage is not available to P2P lending platforms and stablecoin solutions, which face higher risks and regulatory uncertainties.

### References

- Finance Magnates. (2023). The Emergence of P2P Lending Platforms: Disrupting Traditional Lending. Retrieved from [financemagnates.com](https://www.financemagnates.com/fintech/p2p/the-emergence-of-p2p-lending-platforms-disrupting-traditional-lending-institutions/)
- LiquiLoans. (2023). P2P Lending - The Next Best Alternative to Traditional Loans and Investment. Retrieved from [liquiloans.com](https://blog.liquiloans.com/p2p-lending-the-next-best-alternative-to-traditional-loans-and-investment-141b4b215a87)
- Fintech Basics. (2023). P2P Lending Platforms vs Traditional Banks. Retrieved from [fintechbasics.com](https://fintechbasics.com/p2p-lending-platforms-vs-traditional-banks/)
- Monexo. (2023). P2P Lending vs Traditional Lending. Retrieved from [monexo.co](https://monexo.co/p2p-lending-vs-traditional-lending/)
- IADI. (2019). Private Deposit Insurance, Deposit Flows, and Bank Lending. Retrieved from [iadi.org](https://www.iadi.org/en/assets/File/Special_Workspace/2019_Research_Conference/PAPER%204%20-%20Schaeck%20et%20al.pdf)
- Bank of England. (2023). Regulatory Regime for Systemic Payment Systems Using Stablecoins and Related Service Providers. Retrieved from [bankofengland.co.uk](https://www.bankofengland.co.uk/paper/2023/dp/regulatory-regime-for-systemic-payment-systems-using-stablecoins-and-related-service-providers)
- Statista. (2020). UK: Peer-to-Peer Lending vs Bank Lending 2020. Retrieved from [statista.com](https://www.statista.com/statistics/797749/peer-to-peer-vs-bank-lending-uk/)
- Oxford Academic. (2023). Deposit Insurance Premiums and Bank Risk. Retrieved from [academic.oup.com](https://academic.oup.com/rcfs/article/12/2/291/7035951)

### Impact of Regulatory Fines and Enforcement Actions on Fintech Payment Innovations

**Regulatory fines and enforcement actions** in the cryptocurrency space have a significant impact on the development and adoption of fintech payment innovations. Here are some key reflections:

- **Increased Compliance and Security:** Regulatory scrutiny compels fintech companies to enhance their compliance and security measures, leading to more robust and secure payment systems. This can increase consumer trust and adoption (Cointelegraph, 2024).
- **Innovation Slowdown:** Stringent regulations and hefty fines can stifle innovation. Startups and smaller fintech firms may struggle to meet regulatory requirements, which can slow down the pace of innovation and development in the sector (Springer, 2024).
- **Market Consolidation:** Larger, well-established companies with the resources to comply with regulations may dominate the market, leading to consolidation. This can reduce competition and limit the diversity of solutions available to consumers (CCN, 2024).
- **Global Regulatory Harmonization:** Different countries implement varying regulations, presenting challenges for fintech companies operating internationally. This can hinder the global adoption of innovative payment solutions (Channel News Asia, 2024).
- **Focus on Transparency and Accountability:** Regulatory actions emphasize the need for transparency and accountability in financial transactions, leading to the development of more transparent and user-friendly fintech solutions (Cointelegraph, 2024).

### Do Regulatory Fines Restrain the Growth of Fintech?

**Yes, regulatory fines can restrain the growth of fintech** in several ways:

- **Financial Burden:** Fines and penalties impose a significant financial burden on fintech companies, especially startups. This diverts resources away from innovation and growth towards legal and compliance costs (CCN, 2024).
- **Risk Aversion:** The threat of regulatory fines can make fintech companies more risk-averse. They may avoid exploring new and potentially disruptive technologies due to fear of non-compliance and subsequent penalties (Springer, 2024).
- **Barrier to Entry:** Stringent regulatory requirements and the associated risk of fines create high barriers to entry for new players in the fintech space, limiting the influx of fresh ideas and innovations (CCN, 2024).
- **Operational Challenges:** Constantly evolving regulations require fintech companies to continuously update their compliance frameworks, which can be operationally challenging and costly, particularly for smaller firms (Cointelegraph, 2024).

However, it is also important to note that regulatory actions can have positive effects by:

- **Enhancing Consumer Trust:** By ensuring that fintech companies adhere to high standards of security and transparency, regulatory actions can enhance consumer trust in fintech solutions (Cointelegraph, 2024).
- **Leveling the Playing Field:** Regulations help level the playing field by ensuring that all players, including traditional financial institutions and fintech startups, operate under similar rules (Channel News Asia, 2024).

In summary, while regulatory fines and enforcement actions can pose challenges to the growth of fintech, they also play a crucial role in ensuring the stability, security, and trustworthiness of the financial ecosystem.

### References

- Cointelegraph. (2024). Fines and regulation: The ever-growing landscape of crypto compliance. Retrieved from [cointelegraph.com](https://cointelegraph.com/news/fines-and-regulation-the-ever-growing-landscape-of-crypto-compliance).
- Channel News Asia. (2024). Balancing fintech innovation and regulation in the crypto space. Retrieved from [channelnewsasia.com](https://www.channelnewsasia.com/business/money-mind-cryptocurrency-regulation-2435346).
- CCN. (2024). Crypto & Fintech Firms Hit with $5.8 Billion in Fines. Retrieved from [ccn.com](https://www.ccn.com/news/crypto-fintech-firms-hit-with-5-8-billion-fines-global-clampdown-illegal-funds/).
- Springer. (2024). Cryptocurrencies: Regulatory Perspectives and Implications for Fintech. Retrieved from [link.springer.com](https://link.springer.com/chapter/10.1007/978-3-030-65117-6_7).
- Skadden. (2024). As Blockchain Technology and Cryptocurrency Mature, so Do Their Regulations. Retrieved from [skadden.com](https://www.skadden.com/insights/publications/2021/01/2021-insights/regulatory/as-blockchain-technology-and-cryptocurrency-mature).